feat: add featurestore module including Featurestore, EntityType, and Feature classes; add get, update, delete, list methods in all featurestore classes; add search method in Feature class

morgandu opened the pull request:
feat: add featurestore module including Featurestore, EntityType, and Feature classes; add get, update, delete, list methods in all featurestore classes; add search method in Feature class

diemtvu commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,456 @@
+
+    def __init__(
+        self,
+        entity_type_name: str,
```
Is this a common pattern in other resource? I personally find this a bit confuse for overload this arg with both short form and long form, which make a weird case when full path is used, but project/location/featurestore_id also provided. Is this possible to have overloading constructor in Python?

morgandu replied:
Python does not support overloading constructor, that's why we have this as the common pattern for all other Vertex SDK resource classes, for example this is Vertex AI SDK [Dataset class constructor](https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform/datasets/dataset.py#L61).

Agree on the confusion of the usage of optional arguments, but that's the language limitation we are working with. We try to compensate by emphasizing the Docstring and Example Usage throughout the SDK.

diemtvu commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,456 @@
+    def __init__(
+        self,
+        entity_type_name: str,
+        featurestore_id: Optional[str] = None,
```
it's probably easier to keep the project , location, featurestore_id order

morgandu replied:
One of the usage pattern for Vertex SDK is user to initialize the project, see usage example https://github.com/googleapis/python-aiplatform#initialization. And user does not need to specify `project` and `location` unless overwritten. That's why Vertex SDK normally ranks `project` and `location` to bottom among other arguments and mark as optional.

Also `Optional` arguments can not go before the `Required` arguments.

diemtvu commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,456 @@
+                Example: "projects/123/locations/us-central1/featurestores/my_featurestore_id/entityTypes/my_entity_type_id"
+                or "my_entity_type_id" when project and location are initialized or passed, with featurestore_id passed.
+            featurestore_id (str):
+                Optional. Featurestore to retrieve entityType from.
```
What is the value if not set?

morgandu replied:
The optional `featurestore_id` defaults to `None`, see couple of lines above. 
```
featurestore_id: Optional[str] = None,
```

diemtvu replied:
I mean if it's None, does it mean the value from the entity_type_name name will be used, which implies entity_type_name must be the fully qualified path. And when it's None, then entity_type_name must be the short ID form. Probably adding some explanation to be clear

diemtvu commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,456 @@
+        """
+        return _featurestores.Featurestore(self._featurestore_name)
+
+    def get_feature(self, feature_id: str) -> "_featurestores.Feature":
```
Do we expect the feature resource exist (i.e it was created before)? If not, then should the (Feature) update function throw error in that case?

Similar for get_entity_type for featurestore class.

morgandu replied:
Both get_* methods will pass in the fully-qualified resource name into resource class constructor. If the resource exists, the method will return the constructed resource class, if the resource does not exist, it will actually hit the service client's get_* method and throw service error.

lclc19 commented on google/cloud/aiplatform/utils/featurestore_utils.py:
```diff
@@ -0,0 +1,73 @@
+FEATURE_NAME_PATTERN_REGEX = (
+    ENTITY_TYPE_NAME_PATTERN_REGEX
+    + r"\/features\/(?P<feature_id>"
+    + RESOURCE_ID_PATTERN_REGEX
```
Can you add more check for feature_id -- it cannot be one of reserved words: entity_id, feature_timestamp, arrival_timestamp.

morgandu replied:
Vertex SDK is relying on the service for this type of value validation, most validation we are doing is to assist certain usage within the SDK which is to prepare the value to be passed into the service, i.e. these pattern validations essentially are to assist the limitation due to lack of overloading constructor, and we are validating the arguments and form the final resource name to pass into the service. 

However, if the service does not throw an error on the validation you proposed and will cause un alerted error, we can definitely add these in the SDK layer.

Please let me know!

lclc19 commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,456 @@
+        """Deletes feature resources in this EntityType given their feature IDs.
+        WARNING: This deletion is permanent.
+
+        Args:
```
Add feature_ids description

morgandu replied:
Good catch! Thanks!

morgandu commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,525 @@
+
+    def __init__(
+        self,
+        featurestore_name: str,
```
`project` and `location` are `Optional` arguments, whereas `featurestore_name` is `Required` argument. The required arguments have to be listed before the optional arguments. There was another [discussion](https://github.com/googleapis/python-aiplatform/pull/850#discussion_r752859581) of why our `project` and `location` are marked as optional.

morgandu commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,525 @@
+        ):
+            raise ValueError(
+                f"{featurestore_name} is not in form of a fully-qualified "
+                f"featurestore resource name nor a featurestore ID."
```
Refactored this raise ValueError into `featurestore_utils.py` and it does not involve project and location.

morgandu commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,525 @@
+            ValueError if the provided entity_type_id is not in form of a entity_type ID.
+        """
+        if not featurestore_utils.validate_id(entity_type_id):
+            raise ValueError(f"{entity_type_id} is not in form of a entity_type ID.")
```
Done!

sasha-gitg commented on google/cloud/aiplatform/utils/featurestore_utils.py:
```diff
@@ -0,0 +1,188 @@
+    return bool(re.compile(r"^" + RESOURCE_ID_PATTERN_REGEX + r"$").match(resource_id))
+
+
+def validate_featurestore_name(featurestore_name: str) -> Dict[str, str]:
```
Preference to name these `parse` instead of `validate`.

Also, is additional RESOURCE_ID validation the main reason we are not relying on the GAPIC parse methods[1]?

1. https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform_v1/services/featurestore_service/client.py#L230

morgandu replied:
Yes, I am still evaluating the restrictions on Feature Store's resource ids, since we are enabling resource creation/ingestion through user provided df. My gut feeling is we may need some validation in the SDK layer but I think I will have clearer idea when implementing the df related methods. 

I will use the GAPIC methods for now and revisit later if they are needed.

sasha-gitg commented on google/cloud/aiplatform/utils/featurestore_utils.py:
```diff
@@ -0,0 +1,188 @@
+    return f"featurestores/{featurestore_id}/entityTypes/{entity_type_id}/features"
+
+
+def validate_and_get_featurestore_resource_id(featurestore_name: str,):
```
```suggestion
def validate_and_get_featurestore_resource_id(featurestore_name: str,) -> str:
```

sasha-gitg commented on google/cloud/aiplatform/utils/featurestore_utils.py:
```diff
@@ -0,0 +1,188 @@
+
+def validate_and_get_entity_type_resource_ids(
+    entity_type_name: str, featurestore_id: Optional[str] = None,
+):
```
```suggestion
) -> Tuple[str, str]:
```

sasha-gitg commented on google/cloud/aiplatform/utils/featurestore_utils.py:
```diff
@@ -0,0 +1,188 @@
+    feature_name: str,
+    featurestore_id: Optional[str] = None,
+    entity_type_id: Optional[str] = None,
+):
```
```suggestion
) -> Tuple[str, str, str]:
```

morgandu commented on google/cloud/aiplatform/utils/__init__.py:
```diff
@@ -453,6 +461,23 @@ class EndpointClientWithOverride(ClientWithOverride):
+        (compat.V1, featurestore_service_client_v1.FeaturestoreServiceClient),
+        (compat.V1BETA1, featurestore_service_client_v1beta1.FeaturestoreServiceClient),
+        (
+            "online_" + compat.V1,
```
added an `FeaturestoreOnlineServingClientWithOverride` matching the current pattern

morgandu commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,436 @@
+            List[EntityTypes] - A list of managed entityType resource objects
+        """
+
+        cls._featurestore_id = featurestore_utils.validate_and_get_featurestore_resource_id(
```
I have already changed the base `_list()` method to include an optional `parent` field for `search_feature`. Updated couple of unnecessary `cls._* =*` which is causing the confusion on mutating the class, those are only to parse some ids and parent resource name for retrieving child resources. Let me know if this resolve your concern. 

However, one thing I would need more guidance is the forming of `_resource_noun` for the nested resource, which require the ancestor IDs as part of the _resource_noun, which can not be pre defined.

morgandu commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,525 @@
+        """Deletes entity_type resources in this Featurestre given their entity_type IDs.
+        WARNING: This deletion is permanent.
+
+        Args:
```
Done

morgandu commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,527 @@
+
+from typing import Dict, List, Optional, Sequence, Tuple, Union
+
+import pandas as pd
```
We can replace the 
```import pandas as pd``` 
with
```
    import pandas as pd
except ImportError:
    raise ImportError(...)
 ```
But what we normally do with the type hints of the methods? Placing the import inside the method will require a different type hint?

morgandu commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,527 @@
+        Raises:
+            ValueError if the provided entity_type_id is not in form of an entity_type ID.
+        """
+        if not featurestore_utils.validate_id(entity_type_id):
```
The `featurestore.get_entity_type` requires an `entity_type_id` only that's why there is an explicit `featurestore_utils.validate_id(entity_type_id):` in use. For the `EntityType` constructor, since it requires an `entity_type_name` and an optional `featurestore_id`, the validation is inside the `featurestore_utils.validate_and_get_entity_type_resource_ids`.

sasha-gitg reviewed (commented):
Please add a more informative PR title because it will be added to the release notes.

morgandu commented:
- Addressed feedback to remove parent resource as property, instead utilize `gca_resource` i.e. `self.resource_name` when needed. 
- Cleaned up unit tests according to the changes. 
- Added integration test, though in this PR we can only cover `list()` and `search_features()` methods in `Featurestore` class.

sasha-gitg commented on google/cloud/aiplatform/_featurestores/feature.py:
```diff
@@ -0,0 +1,306 @@
+            featurestore_id=featurestore_id,
+        )
+
+        self._resource_noun = (
```
Add a comment this this is a workaround until the base class supports nested features.

sasha-gitg commented on google/cloud/aiplatform/_featurestores/feature.py:
```diff
@@ -0,0 +1,306 @@
+    @property
+    def featurestore_name(self) -> str:
+        """Full qualified resource name of the managed featurestore in which this Feature is."""
+        return "/".join(self.resource_name.split("/")[:-4])
```
Preference to use GAPIC's parsers here

sasha-gitg commented on google/cloud/aiplatform/_featurestores/feature.py:
```diff
@@ -0,0 +1,306 @@
+    @property
+    def entity_type_name(self) -> str:
+        """Full qualified resource name of the managed entityType in which this Feature is."""
+        return "/".join(self.resource_name.split("/")[:-2])
```
Preference to use GAPIC's parsers here.

sasha-gitg commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,457 @@
+        for entity_type_id in entity_type_ids:
+            entity_type = self.get_entity_type(entity_type_id=entity_type_id)
+            entity_type.delete(sync=sync)
+            if not sync:
```
Same comment here on concurrent deletions.

sasha-gitg commented on google/cloud/aiplatform/_featurestores/entity_type.py:
```diff
@@ -0,0 +1,368 @@
+        for feature_id in feature_ids:
+            feature = self.get_feature(feature_id=feature_id)
+            feature.delete(sync=sync)
+            if not sync:
```
To ensure deletions are as concurrent as possible and reduce the time to delete all features, `wait` should be called after `delete` is invoked on all features. Currently features are deleted sequentially.

sasha-gitg commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,457 @@
+        )
+
+        # Fetch credentials once and re-use for all `_empty_constructor()` calls
+        creds = initializer.global_config.credentials
```
This should default to the credentials passed in then the `global_config` credentials. This is handled by the construction of the resource.

```suggestion
        creds = resource.credentials
```

sasha-gitg commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,457 @@
+        )
+
+        return [
+            _featurestores.Feature(
```
Similar to `_list`, this should use`cls._construct_sdk_resource_from_gapic` to avoid the additional get call on construction.

morgandu replied:
Using `cls._construct_sdk_resource_from_gapic` in `Featurestore.search_features()` will return `Featurestore` class. Move `Featurestore.search_features()` to `Feature.search()` and utilize the class methods to retrieve searched feature list.

sasha-gitg commented on google/cloud/aiplatform/utils/featurestore_utils.py:
```diff
@@ -0,0 +1,149 @@
+    return bool(re.compile(r"^" + RESOURCE_ID_PATTERN_REGEX + r"$").match(resource_id))
+
+
+def validate_and_get_featurestore_resource_id(featurestore_name: str) -> str:
```
This seems to only be used in the `FeatureStore` constructor `EntityType.list`. Both also rely on `utils.full_resource_name` which does similar validation. What is the benefit of using this approach? Is OK to just rely on `utils.full_resource_name`.

sasha-gitg commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,322 @@
+
+    @base.optional_sync()
+    def delete_entity_types(
+        self, entity_type_ids: List[str], sync: Optional[bool] = True,
```
```suggestion
        self, entity_type_ids: List[str], sync: bool = True,
```

sasha-gitg commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,322 @@
+    def delete_entity_types(
+        self, entity_type_ids: List[str], sync: Optional[bool] = True,
+    ) -> None:
+        """Deletes entity_type resources in this Featurestre given their entity_type IDs.
```
```suggestion
        """Deletes entity_type resources in this Featurestore given their entity_type IDs.
```

sasha-gitg commented on google/cloud/aiplatform/_featurestores/featurestore.py:
```diff
@@ -0,0 +1,322 @@
+            featurestore_name=self.resource_name, filter=filter, order_by=order_by,
+        )
+
+    @base.optional_sync()
```
Just realized that this is decorated. I think this is new usage pattern for the SDK and there are three options on usage here to consider. This is worth discussion because this pattern should be used throughout.

**1**:  if this method is decorated then the usage is:

```Python
 feature_store.delete_entity_types(..., sync=False)
 feature_store.wait()  # will throw if any of the deletions had issues
 ```

In this case there is no need to pass `sync` to the underlying deletions and instead we can default to asynchronous to make deletions concurrent. Also note that the method needs to be changed to wait on all entity_type deletes not just the last one as it is currently handled.

```Python 
entity_types = []
for entity_type_id in entity_type_ids:
  entity_type = self.get_entity_type(entity_type_id=entity_type_id)
  entity_type.delete(sync=False)
  entity_types.append(entity_type)

for entity_type in entity_types:
  entity_type.wait()
```

**2**: other option is to pass `sync` directly to the underlying delete calls but not associate them with the FeatureStore:

```Python
 feature_store.delete_entity_types(..., sync=False)
 feature_store.wait()  # will NOT throw if any of the deletions had issues
```

In this case this method should not be decorated with `base.optional_sync`. There is no need to call wait here and the implementation would look like:

```Python 
for entity_type_id in entity_type_ids:
  entity_type = self.get_entity_type(entity_type_id=entity_type_id)
  entity_type.delete(sync=sync)
``` 

**3**: is similar to the second option but instead we can return the list of deleted entity types so the user can wait on each one. So usage look like:

```Python
deleted_entity_types: List[EntityType] = feature_store.delete_entity_types(..., sync=False)
feature_store.wait()  # will NOT throw if any of the deletions had issues

for entity_type in deleted_entity_types:
  entity_type.wait()
```

The implementation would look similar to (2) but we compose the list and return it.

```Python 
entity_types = []
for entity_type_id in entity_type_ids:
  entity_type = self.get_entity_type(entity_type_id=entity_type_id)
  entity_type.delete(sync=sync)
  entity_types.append(entity_type)
return entity_types
``` 

Also to unblock this PR, feel free to remove the async support for now and follow up. Up to you.

morgandu replied:
Thanks for the options. I would vote for option 1 over 2 and 3. The main reason is when we delete child resources from the parent resource, I do want to provide sync option to prevent user to retrieve the delete pending child from the parent, i.e. Featurestore.get_entity_type() when Featurestore.delete_entity_types(). So add a Featurestore.wait() during deleting kind of make sense, and if any of the deletion had issue, it is captured.

sasha-gitg replied:
SGTM

sasha-gitg reviewed (approved):
LGTM! Thanks!
